repo: Add `gofmt` linting

doujiang24 opened the pull request:
Commit Message: 
Additional Description:

just code style fix, make `go fmt` happy.

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

doujiang24 commented:
CI failed due to not allowing `Mixed preceeding tabs and whitespace`:
```
CodeChecker ERROR [glint] Mixed preceeding tabs and whitespace: contrib/golang/filters/http/source/go/pkg/api/capi.go
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=124778&view=logs&j=55bc2ab9-36c1-5ab8-58a6-8c258728eeff&t=33b57bf9-ef13-58c8-bc2b-f804aeb16d39&l=538

Go use tab as indent and use whitespace prefix in multiple line comments.
So, I think skipping this check for Go files does make sense.

But I don't know where to skip this check for Go files.
maybe change `NOGLINT_RE`? in https://github.com/envoyproxy/pytooling/blob/main/envoy.code.check/envoy/code/check/abstract/glint.py#L20

Could someone help to give a suggestion? Thanks!

mattklein123 reviewed (changes requested):
Can you please check this in CI somehow? Perhaps add something to the format CI job?

/wait

doujiang24 commented:
@mattklein123 Thanks. I have added the go code format check in `format_pre.sh`,
also fixed some format things in the old existing go files under `examples`.

But the CI may still fail due to lack of `gofmt`(from Go package) in the CI docker image.
Could you help to update the CI docker image, or if there is a guide to update it?

For skipping checking the go files in the original code check stage, I created a PR, hope it's in the right direction.
https://github.com/envoyproxy/pytooling/pull/627

phlax commented:
> @phlax can provide guidance on the various pieces that need updating. Thank you.

on it just waiting for @doujiang24 's fix to publish and will raise PR envoy-side (its in pytooling rather than build-tools btw)

phlax commented:
`gofmt` workaround can be landed here #24852

doujiang24 commented:
@phlax Thanks, the glint check is passed now.

But the newly added gofmt checking, which is @mattklein123 requested in https://github.com/envoyproxy/envoy/pull/24819#pullrequestreview-1240740357, is complaining the `gofmt` command is missing in the CI docker image, see:
```
find: ‘gofmt’: No such file or directory
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=125016&view=logs&j=55bc2ab9-36c1-5ab8-58a6-8c258728eeff&t=33b57bf9-ef13-58c8-bc2b-f804aeb16d39&l=565

It's introduced here: https://github.com/envoyproxy/envoy/pull/24819/files#diff-bdab1e2a91265fcec2607ff37db74db30c71027adb6a0f044c8b36dc1affdcedR12
That means we are using the `gofmt` command to checking the Go code style.

So, could you help to update the CI docker image, or if there is guidance to update it?

phlax commented:
ah - i see now the reference to build tools - yep that could go in the build image - preferably tho we use bazel - i just had a quick look and couldnt see it provided by `rules_go` or similar but i didnt look too hard

the build tools repo is here https://github.com/envoyproxy/envoy-build-tools

phlax commented:
actually it seems like it is available in rules_go https://github.com/bazelbuild/rules_go/commit/b463dbfef16e07d235800267f1709acf65808396 altho im not sure how it is exposed

doujiang24 commented:
@phlax Thanks.
I think we'd better add golang in the CI image, since we using `gofmt` in the `check_format_go_code.sh`, which doesn't use bazel, just bash instead. And only bash is simpler.

So, I created a PR for envoy-build-tool: https://github.com/envoyproxy/envoy-build-tools/pull/192
Hope it's a proper change, please help to take a look, thanks.

doujiang24 commented:
@phlax another question, `Generate docs` also failed now, without a helpful error message, see:
```
##[error]Bash exited with code '1'.
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=125016&view=logs&j=91e633a5-4907-5da1-5862-92a79fe2387a&t=f29e24e7-a093-5724-80bd-d79975d268cb&l=102

Could you help to give some suggestions? thanks.

phlax commented:
The error message is:

```console
/tmp/tmp7snh3ofi/generated/rst/start/sandboxes/dynamic-configuration-control-plane.rst.rst:168: WARNING: line number spec is out of range(1-184): '174-186'
```

The reason for it is that you made a change to the go file in the dynamic-config sandbox and now the `literalinclude` lines dont match correctly - can you take a look at the current docs page (https://www.envoyproxy.io/docs/envoy/latest/start/sandboxes/dynamic-configuration-control-plane) remains the same - you will need to adjust the line numbers 

> So, I created a PR for envoy-build-tool

is there any way we can access `gofmt` from `rules_go` that really would be preferable

phlax commented:
/docs

doujiang24 commented:
@phlax Thanks, `Generate docs` is passed now.

> access gofmt from rules_go that really would be preferable

After a deep look, I think may need to define a bazel rule, similar to `go_binary` to use `go_sdk`.
But I'm not familiar with bazel, it may take a long time. Could we leave it as a TODO? an issue for it?
We have many other PRs for the Go extension, which depends on this PR, since code style will cause conflict.

Another question, `Verify packages` failed now, with the following message, which seems not helpful.
```
PackagesDistroChecker ERROR [distros] Docker image failed to build envoybuild_ubuntu_jammy:latest None
The command '/bin/sh -c chmod +x /tmp/distrotest.sh && apt-get update && apt-get install -y -qq -o=Dpkg::Use-Pty=0 --no-install-recommends curl devscripts gnupg2 procps sudo && mkdir /usr/share/debsign && gpg --no-default-keyring --keyring /usr/share/debsign/keyring.gpg --import /tmp/gpg/signing.key' returned a non-zero code: 100
PackagesDistroChecker INFO [ubuntu_jammy] Testing package: /tmp/tmp3tvpexb8/packages/deb/envoy-1.25_1.25.0-dev_arm64.jammy.changes
PackagesDistroChecker NOTICE [ubuntu_jammy] Building image
```
https://dev.azure.com/cncf/envoy/_build/results?buildId=125241&view=logs&j=de8454aa-2ea9-5422-9f4b-1f3899ba6707&t=d285e0b0-bdfb-593c-c50f-c31c5e3531c1&l=166
Could you help to give some suggestions again? thanks.

phlax commented:
> Could we leave it as a TODO?

really preferably not - adding it to the build image bloats the image and means that you need to use a 5GB image in order to run `gofmt` (or have the correct gofmt in your env)

im pretty sure that we can do this easily - so lets tackle it - i can look but not immediately

 > failed now, with the following message, which seems not helpful.

looks like a network flake

/retest

doujiang24 commented:
@phlax Thanks, I got it.
> need to use a 5GB image in order to run gofmt

Or, could we leave the go format checking thing in CI as a TODO?
This PR really fixed the code style, only lacking a CI check. seems to make sense?

We do not want the code style thing blocking the new features, we do have many other features for the Go extension.
Thoughts? @phlax @mattklein123 Thanks.

mattklein123 commented:
> We do not want the code style thing blocking the new features, we do have many other features for the Go extension.
Thoughts? @phlax @mattklein123 Thanks.

No, let's please get this fixed, thanks.

doujiang24 commented:
@mattklein123 @phlax 
Go code style checking has been added, and the failed test `zstd example` seems doesn't relate to these changes.
Please help take another look, thanks.

phlax commented:
@doujiang24 the error looks genuine ...

```console
Step 6/7 : RUN cd go-control-plane && make bin/example
 ---> Running in 75b3226fdbe6
go: downloading github.com/golang/protobuf v1.4.3
go: downloading google.golang.org/grpc v1.36.0
go: downloading github.com/cncf/xds/go v0.0.0-20210312221358-fbca930ec8ed
go: downloading github.com/envoyproxy/protoc-gen-validate v0.1.0
go: downloading google.golang.org/protobuf v1.25.0
go: downloading google.golang.org/genproto v0.0.0-20200526211855-cb27e3aa2013
go: downloading golang.org/x/net v0.0.0-20200822124328-c89045814202
go: downloading github.com/census-instrumentation/opencensus-proto v0.2.1
go: downloading golang.org/x/sys v0.0.0-20200323222414-85ca7c5b95cd
go: downloading golang.org/x/text v0.3.0
# github.com/envoyproxy/go-control-plane/internal/example
internal/example/resource.go:157:30: undefined: resource
internal/example/resource.go:160:31: undefined: resource
internal/example/resource.go:173:25: undefined: cache
internal/example/resource.go:174:9: undefined: cache
make: *** [Makefile:79: bin/example] Error 2
The command '/bin/sh -c cd go-control-plane && make bin/example' returned a non-zero code: 2
Service 'go-control-plane' failed to build : Build failed
```

could you check the go imports that have changed in `resource.go`

im wondering why the imports are removed - is `gofmt` showing them as unused ?

doujiang24 commented:
@phlax Wow, thanks, I see. I have added back the import packages in `resource.go`, with `_` prefix, making them anonymous.

They were removed by `goimport` by my `goland` editor, which enabled the `goimport` auto-formatting, not by `gofmt`.

phlax commented:
cool re imports

re `gofmt` <> bazel - it seems like it is already there

```console
$ bazel run @go_sdk//:bin/gofmt -- --help
...snip...
usage: gofmt [flags] [path ...]
  -cpuprofile string
        write cpu profile to this file
  -d    display diffs instead of rewriting files
  -e    report all errors (not just the first 10 on different lines)
  -l    list files whose formatting differs from gofmt's
  -r string
        rewrite rule (e.g., 'a[b:len(a)] -> a[b:]')
  -s    simplify code
  -w    write result to (source) file instead of stdout
```

doujiang24 commented:
@phlax Wow, thanks.
I have changed to use `bazel run @go_sdk//:bin/gofmt` instead of introducing a new bazel rule, which makes code much simpler.

phlax commented on tools/code_format/check_format_go_code.sh:
```diff
@@ -0,0 +1,20 @@
+
+cd "$root" || exit 1
+
+gofmt="@go_sdk//:bin/gofmt"
```
```suggestion
GOFMT_BIN="bazel run @go_sdk//:bin/gofmt -- "
```
if the intention is to make the gofmt configurable in some way then probably use a bash global explicitly and the full bazel command so that you can override it without bazel

phlax reviewed (commented):
thanks for working on this @doujiang24 - i think overall this is a very welcome addition

im not a go dev so was unsure about how best to set `gofmt` up - eg i was wondering about tabs/spaces

i asked chatGPT - which said:

> It is more common to use tabs with gofmt, the official Go formatter. Go's style guidelines recommend using tabs for indentation and gofmt enforces this by default. However, spaces can also be used with gofmt by using the -s flag to simplify the code.

so happy to go with the flow on that - we can always change our formatting style after if we decide that is what we want

ideally it would be good to incorporate this into `envoy.code.check` which can then do file discovery, error/warning handling, parallel running etc - but i think that can happen after - just getting the code formatted and a check to keep it that way is a great first step

phlax commented on tools/code_format/check_format_go_code.sh:
```diff
@@ -0,0 +1,20 @@
+files=$(find . -name "*.go" -exec bazel run $gofmt -- -l {} \;)
+if [[ $files != "" ]]; then
+  # write changes to original files, so that we can get the changes by git diff.
+  find . -name "*.go" -exec bazel run $gofmt -- -w {} \;
```
would it be possible to run this once with a list of files ?

currently its a bit of a log hog in CI as it runs each as a separate bazel job

relatedly its triggering a "WARNING ... @go_sdk//:bin/gofmt is a source file, nothing will be built for it. If you want to build ..." on each invokation

not sure exactly what is causing this - i get the same if i try to build it locally - seems like an upstream issue

either way - if we can just invoke the bin once that will limit the output

phlax replied:
```console

In tools/code_format/check_format_go_code.sh line 19:
  $GOFMT_BIN -w $go_files
                ^-------^ SC2086 (info): Double quote to prevent globbing and word splitting.

```

i think you will need to use a bash `${ARRAY[@]}` for `go_files`

phlax commented on tools/code_format/check_format_go_code.sh:
```diff
@@ -0,0 +1,23 @@
+
+set -e
+
+GOFMT_BIN="bazel run @go_sdk//:bin/gofmt -- "
```
```suggestion
GOFMT_BIN="${GOFMT_BIN:-bazel run @go_sdk//:bin/gofmt -- }"
```

phlax replied:
i think lets add this - its required you to use your host `gofmt`

phlax commented on examples/dynamic-config-cp/resource.go:
```diff
@@ -27,8 +27,8 @@ import (
-	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
-	"github.com/envoyproxy/go-control-plane/pkg/resource/v3"
+	_ "github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	_ "github.com/envoyproxy/go-control-plane/pkg/resource/v3"
```
i think this still fails CI (apologies if you fixed somehow)

if it is - then i think just revert it and only worry about `gofmt` linting in this PR

doujiang24 replied:
Yeah, thanks, I also noticed this failure. I have fixed it in a new commit, using their package names.

I tested it locally, and the go-control-plane compiles ok now.

phlax commented on tools/code_format/check_format_go_code.sh:
```diff
@@ -13,10 +13,10 @@ cd "$root" || exit 1
 go_files=$(find . -name "*.go")
```
arggh - maybe not 

> The problem is that the shell performs word splitting on the results of find so that the elements of the array are not guaranteed to be what you want.

https://stackoverflow.com/questions/23356779/how-can-i-store-the-find-command-results-as-an-array-in-bash

chatGPT says:

```bash

#!/bin/bash

IFS=$'\n'
find /path/to/search -type f -print0 | while read -d $'\0' file; do
    echo "Processing $file"
    # Add any additional commands here
done

```

phlax replied:
...

```bash

#!/bin/bash

go_files=()

IFS=$'\n'
find /path/to/search -type f -print0 | while read -d $'\0' file; do
    go_files=+("$file")
done

```

phlax replied:
...

```bash
#!/bin/bash

go_files=()

IFS=$'\n'
while IFS= read -r -d $'\0' file; do
    go_files+=("$file")
done < <(find /path/to/search -type f -print0)


```

phlax replied:
i think you need this ^^ @doujiang24

doujiang24 replied:
Oh, I added parentheses. And I tested it locally, it works, even without parentheses.
I think it's good enough for us. We do not use space in filenames.

Maybe we can keep it simple? I'm fine to change it if you think `while read` could be better.

phlax replied:
> I think it's good enough for us. 

lets do it the right way

doujiang24 replied:
Thanks, already fixed it, also tested the shell works on my local side.

phlax reviewed (commented):
just the one last nit and i think we are gtm - thanks for your hard work on this @doujiang24

doujiang24 commented:
@phlax Thanks for your patience and help~ Tests all passed~

phlax reviewed (approved):
lgtm, thanks again @doujiang24
